File: src/leetcode/p1679/Solution.java
package leetcode.p1679;
// https://leetcode.com/problems/max-number-of-k-sum-pairs/

import java.util.Arrays;

public class Solution {
    public int maxOperations(int[] nums, int k) {
        int n= nums.length;
        if (n == 1) return 0;

        Arrays.sort(nums);
        int l= 0, r= n-1, ans= 0;
        while (l < r) {
            if (nums[r]+nums[l] > k) {
                --r;
                continue;
            }
            if (nums[l]+nums[r] < k) {
                ++l;
                continue;
            }
            if (nums[l]+nums[r] == k) {
                ++ans;
                ++l;
                --r;
            }
        }
        return ans;
    }
}

File: src/leetcode/cpp/p1679/solution.cpp
// https://leetcode.com/problems/max-number-of-k-sum-pairs/?envType=study-plan-v2&envId=leetcode-75
#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;

/**
 * Two Pointers
 * Time: O(nlog(n))
 * Space: O(1)
 */
class Solution {
public:
    int maxOperations(vi& nums, int k) {
        int n= nums.size();
        sort(nums.begin(), nums.end());
        int ans= 0, l= 0, r= n-1;
        while (l < r) {
            if (nums[l] + nums[r] == k) {
                ++ans;
                ++l;
                --r;
            } else if (nums[l] + nums[r] < k) {
                ++l;
            } else {
                --r;
            }
        }
        return ans;
    }
};